prov/efa/src/rxr/rxr_pkt_type_misc.c

```diff
@@ -63,45 +63,46 @@ ssize_t rxr_pkt_init_handshake(struct rxr_ep *ep,
 
 	pkt_entry->pkt_size = sizeof(struct rxr_handshake_hdr)
 			      + RXR_NUM_PROTOCOL_VERSION * sizeof(uint64_t);
-	pkt_entry->addr = addr;
+	pkt_entry->addr = rx_entry->addr;
+	pkt_entry->x_entry = rx_entry;
 	return 0;
 }
 
-void rxr_pkt_post_handshake(struct rxr_ep *ep,
-			    struct rxr_peer *peer,
-			    fi_addr_t addr)
+ssize_t rxr_pkt_post_handshake(struct rxr_ep *ep, struct rxr_peer *peer,
+				fi_addr_t addr)
 {
-	struct rxr_pkt_entry *pkt_entry;
+	struct rxr_rx_entry *rx_entry;
 	ssize_t ret;
 
 	assert(!(peer->flags & RXR_PEER_HANDSHAKE_SENT));
 
-	pkt_entry = rxr_pkt_entry_alloc(ep, ep->tx_pkt_efa_pool);
-	if (OFI_UNLIKELY(!pkt_entry))
-		return;
+	rx_entry = ofi_buf_alloc(ep->rx_entry_pool);
+	if (OFI_UNLIKELY(!rx_entry)) {
```

Comment:
> > However, if posting handshake packet fails at peer side, the sender will keep waiting, which results in hanging issue.
> 
> Was this because no additional packets were received from the sender, so `rxr_pkt_post_handshake` was never called again to send the packet to the sender? Or was it something else?

Correct, the sender sets the flag `RXR_PEER_REQ_SENT` after sending the REQ packet, so the sender just waits for handshake packet and it won't send additonal packets to the peer.

Reply:
> I'm not sure if this is the best approach here, now we're going to crash if we fail to allocate an `rx_entry` when this is called. Given we're going to allocate these for every REQ packet received during startup I think that could be a problem.
> 
> I'd prefer something else, maybe a list of peers that we failed to send handshake packets for. Those peer structures are already allocated by the time this function is called. This approach also lets us prioritize sending these packets over others (like segments, etc.) in the progress function.

That's also the reason we switched to use rx_entry pool here, as it can grow.
I agree with you here, and actually thought about something similar. I was thinking to scan all peers in progress engine and retry/repost those peers that we failed to allocate rx_entry for, but felt that's too much cost in the critical path. 
As you sugguested, we probably can add another list struct and just link those failed peer (failed to allocate rx_entry for), and retry/repost when entering progress engine. 

Next step, I plan to address all minor comments in this commit, and have a separate commit in this PR for the peer list that needs to be retried. Let me know if that's acceptable.

Reply:
I think use a separate commit to implement peer list is a good idea. We will need to evaluate its performance impact. Adding the peer list means we will be looping over that list in progress engine every time. Even though that list will be mostly empty after initialization, there are still costs associated with looping over it.

Reply:
Why is the rx_entry needed at all?

The commits should be add peer list + implement this commit, not this commit and then add the peer list, though. And if each commit doesn't stand on its own (compiles and works independently) then just do one.

I'm not worried about checking an empty list and don't think you need to evaluate the impact. I'll take the cache miss for checking two list pointers over trying to obtain an rx_entry in this path.

Reply:
In this case `rx_entry` is only used to hold addr, nothing else. If we are not concerned the performance impact, we can do it in two commits. First one introduce queued_peer_list in rxr_ep, and loop over it in progress engine. Next one put the peer in queue_peer_list if posting handshake encountered EAGAIN. In that case, there is not must point of reusing `rxr_pkt_post_ctrl_or_queue`. We might just have `rxr_pkt_post_handshake_or_queue`.

Reply:
Alright, I think i got the basic idea here. Once EAGAIN happens when posting handshake packet, we just add that peer into the queue_peer_list and retry later. During retry, we will reconstruct the pkt_entry. So rx_entry is not needed. The work needed is basically as follows:
1. Inside rxr_pkt_handle_recv_completion(), call rxr_pkt_post_handshake_or_queue(). If an error other than EAGAIN returns, hard fail.
2. Inside rxr_pkt_post_handshake_or_queue(), if EAGAIN happens, add that peer into queue_peer_list for retry later.
3. Inside rxr_ep_progress_internal, add a loop to scan queue_peer_list, call rxr_pkt_post_handshake_or_queue for each peer.

Reply:
> Inside rxr_ep_progress_internal, add a loop to scan queue_peer_list, call rxr_pkt_post_handshake_or_queue for each peer.

You do not want to call `rxr_pkt_post_handshake_or_queue` inside progress engine, because it is already queued. You need a separate function `rxr_pkt_post_handshake` to just post handshake. Both `rxr_pkt_post_handshake_or_queue` and progress engine call this function.

Reply:
That's what I had in mind, yeah. Another option is to trigger the handshake during av_insert which could potentially move this traffic to job init/startup time instead of on the first packet. But I think that should be explored separately.

Reply:
Okay, then I will put these change in the current commit for simplicity.

Reply:
Updated in new revision.